[core] Introduce core chaos network release tests

Sparks0219 opened the pull request:
> Briefly describe what this PR accomplishes and why it's needed.

Creating core chaos network release tests by adding ip table variations to the current chaos release tests. Also added a basic chaos release test for streaming generators and object ref borrowing. 

Did a minor refactor by moving each chaos test workload (tasks/actors/streaming gen/borrowing) into it's own python file so it's easier to add additional tests in the future rather than making a huge mono file. Added metrics for total runtime + peak head node memory usage. Furthermore removed the baseline run as it's repeated among all chaos failure types, and it should only be run once in total. Hence for each workload, we now have 4 tests (baseline, EC2 instance killer, raylet killer, ip table network failure). 

Note that for the ip table tests you'll need to add these 4 config variables:
        - RAY_health_check_period_ms=10000
        - RAY_health_check_timeout_ms=100000
        - RAY_health_check_failure_threshold=10
        - RAY_gcs_rpc_server_connect_timeout_s=60
 the top 3 prevent the raylet from failing the gcs health check during the transient network error duration, and the last prevents us from getting killed by the GCS client check where upon connection if we can't initially connect to the GCS for 5 seconds, we die. 
 
 Also deleted test_chaos.py that's located in python/ray/tests as the release chaos tests cover similar functionality.

Sparks0219 committed:
Introduce core chaos network release tests + streaming gen and borrowing tests

Signed-off-by: joshlee <joshlee@anyscale.com>

Sparks0219 committed:
Fix pytest errors

Signed-off-by: joshlee <joshlee@anyscale.com>

Sparks0219 committed:
Fixing tests

Signed-off-by: joshlee <joshlee@anyscale.com>

Sparks0219 committed:
Fixing python tests

Signed-off-by: joshlee <joshlee@anyscale.com>

Sparks0219 committed:
Cleaning up python tests

Signed-off-by: joshlee <joshlee@anyscale.com>

edoakes reviewed (commented):
Stacking all of these tests into one mega-script is a little ugly... is there some lightweight refactoring we can do to clean it up? Like move out shared utils and put each test in its own file. If you don't think this is worthwhile, that's fine.

also, are there any structured metrics we should be tracking for these tests beyond pass/fail?

Sparks0219 committed:
Addressing comments + refactor

Signed-off-by: joshlee <joshlee@anyscale.com>

Sparks0219 commented on release/release_tests.yaml:
```diff
@@ -3413,6 +3444,203 @@
+        - RAY_health_check_period_ms=10000
+        - RAY_health_check_timeout_ms=100000
+        - RAY_health_check_failure_threshold=10
+        - RAY_gcs_rpc_server_connect_timeout_s=60
```
Top 3 are to disable the heartbeat checks which can cause the node to die when transient network errors occur, the last one is a timeout in the gcs client that causes the process to die if it can't initially connect for X amount of time. The default is 5 seconds. 

I'll make a note of this in the PR.

Sparks0219 commented on release/nightly_tests/chaos_test/test_chaos_basic.py:
```diff
@@ -202,23 +353,38 @@ def main():
+        print(
+            "ResourceKiller not found - assuming external chaos injection "
+            "(e.g., iptables, network failures)"
+        )
```
Sg, added a flag for it.

Sparks0219 commented on release/nightly_tests/chaos_test/test_chaos_basic.py:
Ah RIP thanks for catching that, fixed

Sparks0219 commented:
> Stacking all of these tests into one mega-script is a little ugly... is there some lightweight refactoring we can do to clean it up? Like move out shared utils and put each test in its own file. If you don't think this is worthwhile, that's fine.
> 
> also, are there any structured metrics we should be tracking for these tests beyond pass/fail?

The main test script and argument parsing is the same among all workloads so I kept that as is. I moved each workload python function into it's own file so it's cleaner to extend if people in the future want to add additional tests. 

One thing I noticed though was that for example each failure type runs the workload twice, with the first being the baseline. This is unnecessary though because the EC2 Instance killer and the raylet killer will both run the baseline when it should only be ran twice. Also it's broken for ip tables because I run it the network failure starting from the beginning of the python file execution. So I just run the workload once, and have a separate release test that just runs the workload with no failure injection. 

Also followed up from our conversation and added metrics for head node peak mem usage + total runtime which should give us an indicator if things are going wrong.

Sparks0219 committed:
Update bazel file

Signed-off-by: joshlee <joshlee@anyscale.com>

Sparks0219 committed:
AI comment

Signed-off-by: joshlee <joshlee@anyscale.com>

Sparks0219 commented on release/nightly_tests/chaos_test/test_chaos.py:
```diff
@@ -0,0 +1,106 @@
+    print(f"Runtime when there are many failures: {runtime_s}")
+
+    if node_killer is not None:
+        print(f"Total node failures: {ray.get(node_killer.get_total_killed.remote())}")
```
thanks AI!

Sparks0219 committed:
AI comments

Signed-off-by: joshlee <joshlee@anyscale.com>

Sparks0219 committed:
AI comments

Signed-off-by: joshlee <joshlee@anyscale.com>

Sparks0219 committed:
Shorten iptable injection interval

Signed-off-by: joshlee <joshlee@anyscale.com>

edoakes reviewed (approved):
🚀
